Client/Frontend/Browser/SearchLoader.swift

```diff
@@ -51,30 +58,33 @@ class _SearchLoader<UnusedA, UnusedB>: Loader<Cursor<Site>, SearchViewController
             inProgress = deferred as? Cancellable
 
             deferred.upon() { result in
-                self.inProgress = nil
-
+                DispatchQueue.main.async {
+                    self.inProgress = nil
```

Comment:
I'm not 100% convinced that this is safe.

The `inProgress` `Deferred` can be filled on an arbitrary queue. This main dispatch can occur at an arbitrary point in time. It's possible that `didSet` will fire three times in quick succession, and the three `Deferred`s can be filled interleaved in that sequence, as will this dispatch. Even if all three `Deferred`s are filled in the correct sequence, it's possible that this block will run when `inProgress` has already been replaced by a non-run, non-canceled `Deferred`, right?

So I think here you're trying to achieve one thing: when the last `Deferred` fills, unhook itself.

You can do that in one of two ways.

```
guard let inProgress = self.inProgress where inProgress.isFilled || inProgress.cancelled else {
    // Nothing to do: `inProgress` is still running, and therefore must not be us, so do nothing.
    return
}
self.inProgress = nil
```

or the less clear, but a little shorter (the optional coercion might need work):

```
// If the in-progress Deferred isn't us, don't clear it.
if self.inProgress == deferred {
    self.inProgress = nil
}
```

Reply:
The original code appears wrong in regards to cancel/inProgress:
```
inProgress.upon {
   inProgress = nil // basically allowing for function re-entry (and multiple live DB ops for the same purpose)
    // now the func does more DB work, even though another DB op for the same purpose can start due to inProgress = nil line above
}
```

My patch fixes this behaviour (I believe):
```
inProgress.upon {
    DispatchQueue.main.async {  // pseudocode, these are 2 separate main.async calls in the patch (doesn't matter)
        inProgress = nil 
        if there was a result {
           self.urlBar.setAutocompleteSuggestion()
        }
   }
}
```

Given the above, I see no multiple DB ops as you mention in your comment.

Reply:
`inProgress` doesn't control DB operations; it only controls whether we cancel the one that was last there. Whenever we get a `set`, we'll kick off a new DB operation (on the DB queue). They'll run in order, and they'll run independently of the main queue.

Reply:
> Whenever we get a set, we'll kick off a new DB operation
The top of the function has this which is cancelling an previous DB operation. No?
```
   if let inProgress = inProgress {
                 inProgress.cancel()
                 self.inProgress = nil
}
```

Reply:
The thing I'm concerned about is not that — it's that the `.upon` that runs after each DB operation can overlap and interleave with the main-queue code, and potentially run out of order.

My concern about cancelation is that those `.upon` blocks nil out `inProgress`, which might be a non-canceled one (the last one).

And if they run out of order, or their async `main` blocks run out of order, the UI updates won't correspond to user actions.